plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -1,915 +1,1218 @@
+		packageCurrentPowerConsumption.String():     p.addCurrentPackagePower,
+		packageCurrentDramPowerConsumption.String(): p.addCurrentDramPower,
+		packageThermalDesignPower.String():          p.addThermalDesignPower,
+		packageCPUBaseFrequency.String():            p.addCPUBaseFrequency,
+		packageTurboLimit.String():                  p.addMaxTurboFreqLimits,
 	}
-	p.logOnce["intel_uncore_frequency"] = nil
-	p.readUncoreFreq("initial", socketID, die, acc)
-	p.readUncoreFreq("current", socketID, die, acc)
+
+	return nil
 }
 
-func (p *PowerStat) readUncoreFreq(typeFreq string, socketID string, die string, acc telegraf.Accumulator) {
-	fields := map[string]interface{}{}
-	if typeFreq == "current" {
-		if p.areCoreMetricsEnabled() && p.msr.isMsrLoaded() {
-			p.logOnce[socketID+"msr"] = nil
-			cpuID, err := p.GetCPUIDFromSocketID(socketID)
-			if err != nil {
-				p.Log.Debugf("Error while reading socket ID: %v", err)
-				return
-			}
-			actualUncoreFreq, err := p.msr.readSingleMsr(cpuID, msrUncorePerfStatusString)
-			if err != nil {
-				p.Log.Debugf("Error while reading %s: %v", msrUncorePerfStatusString, err)
-				return
-			}
-			actualUncoreFreq = (actualUncoreFreq & 0x3F) * 100
-			fields["uncore_frequency_mhz_cur"] = actualUncoreFreq
-		} else {
-			err := errors.New("Error while trying to read MSR (probably msr module was not loaded), uncore_frequency_mhz_cur metric will not be collected")
-			if val := p.logOnce[socketID+"msr"]; val == nil || val.Error() != err.Error() {
-				p.Log.Errorf("%v", err)
-				// Remember that specific error occurs for socketID to omit logging next time
-				p.logOnce[socketID+"msr"] = err
-			}
+// parsePackageMetrics takes a slice of package metrics. It ensures all metrics in the slice
+// are supported package-specific, and there are no duplicates. If metrics slice is nil, then
+// default package metrics are provided.
+func parsePackageMetrics(metrics []string) ([]string, error) {
+	if metrics == nil {
+		return []string{
+			packageCurrentPowerConsumption.String(),
+			packageCurrentDramPowerConsumption.String(),
+			packageThermalDesignPower.String(),
+		}, nil
+	}
+
+	for _, m := range metrics {
+		if !isValidPackageMetric(m) {
+			return nil, fmt.Errorf("invalid package metric specified: %q", m)
 		}
 	}
-	initMinFreq, err := p.msr.retrieveUncoreFrequency(socketID, typeFreq, "min", die)
-	if err != nil {
-		p.Log.Errorf("Error while retrieving minimum uncore frequency of the socket %s, err: %v", socketID, err)
-		return
-	}
-	initMaxFreq, err := p.msr.retrieveUncoreFrequency(socketID, typeFreq, "max", die)
-	if err != nil {
-		p.Log.Errorf("Error while retrieving maximum uncore frequency of the socket %s, err: %v", socketID, err)
-		return
-	}
 
-	tags := map[string]string{
-		"package_id": socketID,
-		"type":       typeFreq,
-		"die":        die,
+	if hasDuplicate(metrics) {
+		return nil, errors.New("package metrics contains duplicates")
 	}
```

Comment:
Why not directly doing it in the for-loop above?

Reply:
Also, what is the implication of having items multiple times? I don't think there is a real issue?

Reply:
Hmmm, I can see your point, but I'm not sure I like introducing generics for that convenience... @powersj what do you think?

Reply:
After commit [`98e961c4`](https://github.com/influxdata/telegraf/pull/14363/commits/98e961c4affec0767a1e95f9d3722d082952cc1c), `hasDuplicates` supports 3 types. Additionally it is only called once per type, during initialization.